feat(nuxt): add `useResponseHeader` composable

jelmerdemaat opened the pull request:
### 🔗 Linked issue

Resolves https://github.com/nuxt/nuxt/issues/27075

### 📚 Description

Adds 2 new composables:

- `useResponseHeader`
- `useResponseHeaders`

See referenced issue for details.

I tried my best to cover enough ground to make a draft PR:

1. Proposed code for new composables
2. Written 2 new documentation pages

But I still need some help :) With the following questions:

- I'm not familiar with writing tests, is it expected that I add tests and if so, where do I start?
- Right now I propose to return nothing from the functions, but we could also return `true` if successful and `false` if failed? Any ideas for better return values?
- On naming the functions. The name "use" responseHeader kinda implies a return value, as it implies consuming something... Would it be better to name them `setResponseHeader`? That would be in line with the h3 library but, be out of character for a Nuxt composable.
- Do I put the Nuxt versions above the functions myself, or are these added later?
- Typings: right now I used `string` for both header `name` and `value`. But the h3 packages includes a type `HTTPHeaderName` which seems perfect for `name`. But the other composables about headers also didn't use this type, so I didn't use it. Should I use it? Is this something we could improve for all composables about headers?

jelmerdemaat committed:
feat(composables): start new useResponseHeader and useResponseHeaders composables

jelmerdemaat committed:
docs(composables): document new useResponseHeader and useResponseHeaders composables

danielroe commented:
I agree with you - I think that `useResponseHeader` doesn't make sense as a way of setting response headers. I think it should more likely return a computed value, which allows both getting and setting response headers.

For example:

```ts
const someHeader = useResponseHeader('x-some-header')
console.log(someHeader.value) // existing value
someHeader.value = 'new value'
```

jelmerdemaat commented:
@danielroe Any thoughts on my proposed solution before I start building it?

danielroe commented:
Yes, that looks good to me!

jelmerdemaat committed:
feat(composables): change new useResponseHeader composable to return computed value

jelmerdemaat committed:
Merge branch 'main' into feat/useResponseHeader

jelmerdemaat committed:
docs(composables): update new useResponseHeader examples

jelmerdemaat commented:
@danielroe I have updated the `useResponseHeader` code to reflect my proposal. Also updated the docs for it.

To be honest I am unsure about the added value of the plural `useResponseHeaders` and also unsure about how to implement it. We could also go forward with only the singular form? Or I would be glad to update the code for `useResponseHeaders` if you have any tips about implementation (e.g. with a computed, or only as a setter with an object parameter, or...)

jelmerdemaat committed:
Merge branch 'main' into feat/useResponseHeader

danielroe commented:
I agree, let's drop `useRequestHeaders`...

jelmerdemaat committed:
Merge branch 'main' into feat/useResponseHeader

jelmerdemaat committed:
feat(composables): drop plural "useResponseHeaders"

jelmerdemaat committed:
Merge branch 'main' of github.com:nuxt/nuxt into feat/useResponseHeader

jelmerdemaat commented:
I deleted the plural form. 

How do I proceed? I think I need to do some checks/tasks before merging?

jelmerdemaat committed:
Merge branch 'main' of github.com:nuxt/nuxt into feat/useResponseHeader

jelmerdemaat committed:
feat(composables): add basic test for 'useResponseHeader' composable

jelmerdemaat committed:
Merge branch 'main' of github.com:nuxt/nuxt into feat/useResponseHeader

jelmerdemaat commented:
I added a very basic test. I really don't know much about testing Nuxt, so if further work is required, I need some help.

jelmerdemaat committed:
Merge branch 'main' into feat/useResponseHeader

jelmerdemaat committed:
Merge branch 'main' into feat/useResponseHeader

jelmerdemaat committed:
Merge branch 'main' into feat/useResponseHeader

danielroe commented on packages/nuxt/src/app/composables/ssr.ts:
```diff
@@ -61,6 +61,27 @@ export function setResponseStatus (arg1: H3Event | number | undefined, arg2?: nu
   }
 }
 
+/** @since 3.13.2 */
```
```suggestion
/** @since 3.14.0 */
```

danielroe committed:
chore: update `@since` tag

danielroe committed:
Merge branch 'main' into feat/useResponseHeader

danielroe committed:
refactor: warn when setting `useResponseHeader` on client

jelmerdemaat commented on packages/nuxt/src/app/composables/ssr.ts:
```diff
@@ -61,6 +61,43 @@ export function setResponseStatus (arg1: H3Event | number | undefined, arg2?: nu
+    if (import.meta.dev) {
+      return computed({
+        get: () => undefined,
+        set: () => console.warn('[nuxt] Setting response headers is not supported in the browser.'),
```
I think the warning text needs to be changed here? `import.meta.client` was already checked in the code above. In this scenario the `event` is missing

danielroe replied:
you are quite right!

jelmerdemaat committed:
Merge branch 'main' into feat/useResponseHeader

danielroe committed:
Merge remote-tracking branch 'origin/main' into feat/useResponseHeader

danielroe committed:
fix: do not handle missing event on server

danielroe committed:
docs: note about release version

danielroe committed:
Merge remote-tracking branch 'origin/main' into feat/useResponseHeader
